packages/grafana-ui/src/components/uPlot/Plot.tsx

```diff
@@ -117,12 +141,150 @@ export class UPlotChart extends React.Component<PlotProps, UPlotChartState> {
     }
   }
 
+  handleKeyRelease = (e: React.KeyboardEvent) => {
+    if (!SUPPORTED_KEYS.includes(e.key as SupportedKey)) {
+      return;
+    }
+
+    this.keyboardState.pressedKeys[e.key as SupportedKey] = false;
+
+    if (e.key === ' ') {
+      e.preventDefault();
+      e.stopPropagation();
+
+      // We do this so setSelect hooks get fired, zooming the plot
+      this.state.plot?.setSelect(this.state.plot.select);
+      this.keyboardState.dragStartX = null;
+    }
+  };
+
+  handleKeys = (e: React.KeyboardEvent) => {
+    if (!SUPPORTED_KEYS.includes(e.key as SupportedKey)) {
+      return;
+    }
+
+    e.preventDefault();
+    e.stopPropagation();
+
+    const newKey = !this.keyboardState.pressedKeys[e.key as SupportedKey];
+    if (newKey) {
+      const initiateAnimationLoop = Object.values(this.keyboardState.pressedKeys).every((pressed) => pressed === false);
+      this.keyboardState.pressedKeys[e.key as SupportedKey] = true;
+      this.keyboardState.dragStartX =
+        e.key === ' ' && this.keyboardState.dragStartX === null
+          ? this.state.plot?.cursor.left!
+          : this.keyboardState.dragStartX;
+
+      if (initiateAnimationLoop) {
+        window.requestAnimationFrame(this.handlePressedKeys);
+      }
+    }
+  };
+
+  handlePressedKeys = (time: number) => {
+    const nothingPressed = Object.values(this.keyboardState.pressedKeys).every((pressed) => pressed === false);
+    if (nothingPressed || !this.state.plot) {
+      this.keyboardState.keysLastHandledAt = null;
+      return;
+    }
+
+    const dt = time - (this.keyboardState.keysLastHandledAt ?? time);
+    const dx = dt * PIXELS_PER_MS;
+    let horValue = 0;
+    let vertValue = 0;
+
+    if (this.keyboardState.pressedKeys.ArrowUp) {
+      vertValue -= dx;
+    }
+    if (this.keyboardState.pressedKeys.ArrowDown) {
+      vertValue += dx;
+    }
+    if (this.keyboardState.pressedKeys.ArrowLeft) {
+      horValue -= dx;
+    }
+    if (this.keyboardState.pressedKeys.ArrowRight) {
+      horValue += dx;
+    }
+    if (this.keyboardState.pressedKeys.Shift) {
+      horValue *= SHIFT_MULTIPLIER;
+      vertValue *= SHIFT_MULTIPLIER;
+    }
+
+    this.moveCursor(horValue, vertValue);
+
+    const cursor = this.state.plot.cursor;
+    if (this.keyboardState.pressedKeys[' '] && cursor) {
+      const drawHeight = Number(this.state.plot.over.style.height.slice(0, -2));
+
+      this.state.plot.setSelect(
+        {
+          left: cursor.left! < this.keyboardState.dragStartX! ? cursor.left! : this.keyboardState.dragStartX!,
+          top: 0,
+          width: Math.abs(cursor.left! - (this.keyboardState.dragStartX ?? cursor.left!)),
```

Comment:
Are we able to narrow, or default all these `!` assertions out?

It looks like width could possibly end up as NaN here which probably isnt great?

Reply:
@joshhunt if you can help out with https://github.com/leeoniya/uPlot/issues/595, we can get rid of a _many_ `!` in lots of other places that touch uPlot, too.